src/token-processor/util/metadata-helpers.ts

```diff
@@ -0,0 +1,452 @@
+    if (RawMetadataAttributesCType.Check(metadata.attributes)) {
+      for (const { trait_type, value, display_type } of metadata.attributes) {
+        if (trait_type && value) {
+          attributes.push({
+            trait_type: trait_type,
+            value: JSON.stringify(value),
+            display_type: display_type ?? null,
+          });
+        }
+      }
+    }
+    // Localized properties only override their default. All others have to fall back to default
+    // values.
+    const properties: DbMetadataPropertyInsert[] = defaultInsert?.properties ?? [];
+    if (RawMetadataPropertiesCType.Check(metadata.properties)) {
+      for (const [key, value] of Object.entries(metadata.properties)) {
+        if (key && value) {
+          const defaultProp = properties.find(p => p.name === key);
+          if (defaultProp) {
+            defaultProp.value = JSON.stringify(value);
+          } else {
+            properties.push({
+              name: key,
+              value: JSON.stringify(value),
+            });
+          }
+        }
+      }
+    }
+    inserts.push({
+      metadata: metadataInsert,
+      attributes: attributes,
+      properties: properties,
+    });
+    if (inserts.length === 1) {
+      defaultInsert = inserts[0];
+    }
+  }
+  return inserts;
+}
+
+/**
+ * Fetches metadata while monitoring timeout and size limits. Throws if any is reached.
+ * Taken from https://github.com/node-fetch/node-fetch/issues/1149#issuecomment-840416752
+ * @param httpUrl - URL to fetch
+ * @returns JSON result string
+ */
+export async function performSizeAndTimeLimitedMetadataFetch(
+  httpUrl: URL
+): Promise<string | undefined> {
+  const ctrl = new AbortController();
+  let abortReason: Error | undefined;
+
+  const timer = setTimeout(() => {
+    abortReason = new MetadataTimeoutError();
+    ctrl.abort();
+  }, ENV.METADATA_FETCH_TIMEOUT_MS);
+  try {
+    const networkResult = await fetch(httpUrl.toString(), {
+      method: 'GET',
+      signal: ctrl.signal,
+    });
+    if (networkResult.body) {
+      const decoder = new TextDecoder();
+      let responseText: string = '';
+      let bytesWritten = 0;
+      const reportedContentLength = Number(networkResult.headers.get('content-length') ?? 0);
+      if (reportedContentLength > ENV.METADATA_MAX_PAYLOAD_BYTE_SIZE) {
+        abortReason = new MetadataSizeExceededError();
+        ctrl.abort();
+      }
+      for await (const chunk of networkResult.body) {
+        bytesWritten += chunk.byteLength;
+        if (bytesWritten > ENV.METADATA_MAX_PAYLOAD_BYTE_SIZE) {
+          abortReason = new MetadataSizeExceededError();
+          ctrl.abort();
+        }
+        responseText += decoder.decode(chunk as ArrayBuffer, { stream: true });
+      }
+      responseText += decoder.decode(); // flush the remaining bytes
+      clearTimeout(timer);
+      return responseText;
+    }
+  } catch (error) {
+    clearTimeout(timer);
```

Comment:
Similar comment here about http request errors tending to lack useful debug info (request url, http status code, etc)

Reply:
Done